File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutUnsavedChangesRules.kt
package com.example.kpkn.screens.workout

internal sealed interface WorkoutPendingSetAction {
    data class Navigate(val setIdx: Int) : WorkoutPendingSetAction
    data class Edit(val setIdx: Int, val side: String? = null) : WorkoutPendingSetAction
}

internal fun pendingSetNavigationAction(
    hasPendingDraftChanges: Boolean,
    activeSetIdx: Int,
    targetSetIdx: Int,
): WorkoutPendingSetAction? = when {
    targetSetIdx == activeSetIdx -> null
    hasPendingDraftChanges -> WorkoutPendingSetAction.Navigate(targetSetIdx)
    else -> null
}

internal fun pendingSetEditAction(
    hasPendingDraftChanges: Boolean,
    isAlreadyEditingCurrentSet: Boolean,
    targetSetIdx: Int,
    side: String? = null,
): WorkoutPendingSetAction? = when {
    isAlreadyEditingCurrentSet -> null
    hasPendingDraftChanges -> WorkoutPendingSetAction.Edit(targetSetIdx, side)
    else -> null
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutUnsavedChangesRules.swift
import Foundation

enum WorkoutPendingSetAction {
    case navigate(setIdx: Int)
    case edit(setIdx: Int, side: String?)
}

func pendingSetNavigationAction(
    hasPendingDraftChanges: Bool,
    activeSetIdx: Int,
    targetSetIdx: Int
) -> WorkoutPendingSetAction? {
    guard targetSetIdx != activeSetIdx else { return nil }
    return hasPendingDraftChanges ? .navigate(setIdx: targetSetIdx) : nil
}

func pendingSetEditAction(
    hasPendingDraftChanges: Bool,
    isAlreadyEditingCurrentSet: Bool,
    targetSetIdx: Int,
    side: String? = nil
) -> WorkoutPendingSetAction? {
    guard !isAlreadyEditingCurrentSet else { return nil }
    return hasPendingDraftChanges ? .edit(setIdx: targetSetIdx, side: side) : nil
}
